File: Park/CodeTree/String Matching/최장 공통 부분 수열.py
str1 = input()
str2 = input()

str1_len, str2_len = len(str1), len(str2)

# 인덱스를 1부터 시작하게 하기 위해 앞에 '#' 추가
str1, str2 = '#' + str1, '#' + str2

# dp 테이블 초기화
dp = [[0] * (str2_len + 1) for _ in range(str1_len + 1)]

def initialize():
    # 첫 번째 문자의 초기값 설정
    dp[1][1] = int(str1[1] == str2[1])
    
    # 첫 번째 문자열의 첫 문자와 비교하여 초기값 설정
    for i in range(2, str1_len + 1):
        dp[i][1] = 1 if str1[i] == str2[1] else dp[i-1][1]
    
    # 두 번째 문자열의 첫 문자와 비교하여 초기값 설정
    for j in range(2, str2_len + 1):
        dp[1][j] = 1 if str1[1] == str2[j] else dp[1][j-1]

initialize()

for i in range(2, str1_len + 1):
    for j in range(2, str2_len + 1):
        # 문자가 같으면 이전 dp 값에 1을 더함
        if str1[i] == str2[j]:
            dp[i][j] = dp[i-1][j-1] + 1
        # 문자가 다르면 이전 dp 값 중 큰 값을 사용
        else:
            dp[i][j] = max(dp[i-1][j], dp[i][j-1])

print(dp[str1_len][str2_len])

File: jhyun/DP2/String Matching/최장_공통_부분_수열.java
import java.util.*;
import java.io.*;

public class Main {
    static final int MAX_LEN = 1000;

    static String str1, str2;
    static int[][] dp = new int[MAX_LEN + 1][MAX_LEN + 1];

    static int str1Len, str2Len;

    static void init(){
        // dp[1][1] 값은 첫 번째 문자열의 첫 번째 문자와 두 번째 문자열의 첫 번째 문자가 같은지 여부를 저장
        dp[1][1] = str1.charAt(1) == str2.charAt(1) ? 1 : 0;

        // 두 번째 문자열의 1번 인덱스의 문자까지만 사용했을 때 가능한 부분 수열의 최대 길이를 채워넣어줍니다.
        for(int i =2; i <= str1Len; i++){
            if(str1.charAt(i) == str2.charAt(1)){
                dp[i][1] = 1;
            }else{
                dp[i][1] = dp[i-1][1];
            }
        }

        // 첫 번째 문자열의 1번 인덱스의 문자까지만 사용했을 때 가능한 부분 수열의 최대 길이를 채워넣어줍니다.
        for(int j = 2; j <= str2Len; j++){
            if(str1.charAt(1) == str2.charAt(j)){
                dp[1][j] = 1;
            }else {
                dp[1][j] = dp[1][j-1];
            }
        }
    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        str1 = br.readLine();
        str2 = br.readLine();

        str1Len = str1.length();
        str2Len = str2.length();

        // string의 index가 0부터 시작하기 때문에 이를 1부터 시작하기 위해서 앞에 # 추가
        str1 = "#" + str1;
        str2 = "#" + str2;

        init();

        for(int i = 2; i <= str1Len; i++) {
            //첫 번째 문자열의 i 번째까지 문자열을 고려했고 두 변째 문자열의 j 번째까지 문자열을 고려했을 때
            //가능한 부분 수열의 최대 길이 구하기
            for(int j = 2; j <= str2Len; j++){
                if(str1.charAt(i) == str2.charAt(j)){
                    dp[i][j] = dp[i-1][j-1] + 1;
                } else{
                    dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
                }
            }
        }
        System.out.println(dp[str1Len][str2Len]);
    }
}